v4/src/EditNodeModal.jsx

```diff
@@ -0,0 +1,240 @@
+import {
+  Button, Modal, Form, ToggleButtonGroup, ToggleButton,
+} from 'react-bootstrap';
+
+import { PropTypes } from 'prop-types';
+import { NTYPE, RTYPE } from './constants';
+
+import './styles/AddNodeModal.css';
+
+class EditNodeModal extends Component {
+  constructor(props) {
+    super(props);
+    const { addNode, node } = this.props;
+
+    this.state = {
+      isOpen: false,
+      topicName: '',
+      resourceType: (node ? node.data('resource_type') : 0),
+      resourceData: {},
+    };
+
+    this.addNode = addNode;
+    this.toggleModal = this.toggleModal.bind(this);
+    this.handleClose = this.handleClose.bind(this);
+    this.handleOpen = this.handleOpen.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.setResourceType = this.setResourceType.bind(this);
+  }
+
+  setResourceType(resourceType) {
+    this.setState({ resourceType });
+  }
+
+  toggleModal() {
+    const { isOpen } = this.state;
+    this.setState({ isOpen: !isOpen });
+  }
+
+  handleOpen() {
+    const { node } = this.props;
+    this.setState({ // make it default again
+      isOpen: true,
+      topicName: (node ? node.data('name') : ''),
+      resourceType: (node ? node.data('resource_type') : 0),
+      resourceData: (node ? node.data('data') : {}),
+    });
+  }
+
+  handleClose() {
+    this.setState({ // make it default again
+      isOpen: false,
+    });
+    this.toggleModal();
+  }
+
+  handleSubmit() {
+    this.toggleModal();
+    const {
+      node, addNode, setNode, updateEdges, removeNode,
+    } = this.props;
+    const data = { ...node.data() };
+    const { resourceType, resourceData, topicName } = this.state;
+    data.resource_type = resourceType;
+    data.data = resourceData;
+    if (topicName) {
+      data.name = topicName;
+    }
+    if (JSON.stringify(data) === JSON.stringify(node.data())) {
+      return; // No changes to save
+    }
+
+    // if this is a resource node delete its name
+    if (data.node_type === 2) {
+      delete data.name;
+    }
+
+    // if nodetype, rename
+
+    delete data.id;
+
+    const newNode = addNode(data);
+    setNode(newNode);
+
+    // take existing edges and put newNode
+    updateEdges(node, newNode);
+    newNode.shift(node.position());
+    newNode.trigger('tap');
+
+    // delete n
+    removeNode(node);
+  }
+
+
+  render() {
+    const { node } = this.props;
+    const {
+      isOpen, topicName, resourceType, resourceData,
+    } = this.state;
+    if (!node) return <span />;
+    return (
+      <span>
+        <Button className="btn btn-info btn-lg btn-mod" onClick={this.toggleModal}>Edit node</Button>
+        <Modal className="Modal" show={isOpen} onHide={this.handleClose} onShow={this.handleOpen}>
+          <Modal.Header closeButton>
+            <Modal.Title>Edit Node</Modal.Title>
+          </Modal.Header>
+
+          <Modal.Body>
+            <p>{node.data('name')}</p>
+            {node.data('node_type') === NTYPE.TOPIC
```

Comment:
same comment as above. this would be a good time to remove this messy ternary :)